frontend/src/components/UserContext.tsx

```diff
@@ -1,31 +1,102 @@
-import React, { createContext, useState, ReactNode } from 'react';
+import { createContext, ReactNode, useContext, useEffect, useState } from "react";
+import { useNavigate } from "react-router-dom";
 
-interface User {
-  loggedIn: boolean;
-  isAdmin: boolean;
-}
+import { jwtDecode } from 'jwt-decode';
+import { Response } from '../services/apiClient';
+import { Credentials, Token, useLogin, useSignUp } from "../hooks/useUser";
 
 interface UserContextType {
-  user: User | null;
-  loginAsUser: () => void;
-  loginAsAdmin: () => void;
+  token?: string;
+  username?: string;
+  login: (credentials: Credentials) => Promise<Response<Token>>;
   logout: () => void;
+  signUp: (credentials: Credentials) => Promise<Response<string>>;
+  requireLogin: (role: string) => void;
+  role?: string;
+}
+interface JwtToken {
+  sub: string;
+  iat: number;
+  exp: number;
+  iss?: string;
+  aud?: string | string[];
+  role?: string;
+  name: string;
 }
 
+// Create the UserContext
 const UserContext = createContext<UserContextType | undefined>(undefined);
 
-const UserProvider: React.FC<{ children: ReactNode }> = ({ children }) => {
-  const [user, setUser] = useState<User | null>(null);
+// UserContext provider component
+export const UserProvider = ({ children }: { children: ReactNode }) => {
+  const [role, setRole] = useState<string>();
+  const [token, setToken] = useState<string>(localStorage.getItem("token") || "");
+  const [username, setUsername] = useState<string>();
+  const navigate = useNavigate();
+
+  useEffect(() => {
+    if (token) {
+      const decodedToken = jwtDecode<JwtToken>(token);
+      setRole(decodedToken.role?.toLowerCase());
+      setUsername(decodedToken.name);
+    }
+  }, [token]);
 
-  const loginAsUser = () => setUser({ loggedIn: true, isAdmin: false });
-  const loginAsAdmin = () => setUser({ loggedIn: true, isAdmin: true });
-  const logout = () => setUser(null);
+  const login = async (credentials: Credentials): Promise<Response<Token>> => {
+    const loggedInUser = await useLogin(credentials);
 
+    if (loggedInUser && loggedInUser.data) {
+      setToken(loggedInUser.data.token)
+      const decodedToken = jwtDecode(loggedInUser.data.token) as JwtToken;
+      if (decodedToken.role) {
+        localStorage.setItem("token", loggedInUser.data.token);
+        setRole(decodedToken.role.toLowerCase());
+        setUsername(decodedToken.name);
+        navigate('/menu');
+      }
+    }
+    else {
+      console.error('Login failed: result is undefined');
+    }
+    return loggedInUser;
+  };
+
+  const logout = () => {
```

Comment:
For linje 66-68.
Kunne det give mening at samle alle disse variabler i et state eller object der håndtere dette uden for denne komponent?
Kan sagtens være jeg tager fejl og dette fungere fint, har ikke lige det fulde overblik hvordan flowet er her. For mig ligner det et interface der "lever" i UserContext komponenten (Altså oppe i starten af komponenten, linje 8).
Jeg tænker det samme for JWT der starter på linje 17.

Reply:
Ill be honest, im kinda retarded. Jeg tror måske det bliver sat i forbindelsen med useEffect på linje 37, men vil også bare sikre mig at de blev fjernet. Lad mig lige teste om det giver mening at de er der. 

Ift til state, er jeg ikke selv bekendt med hvordan det skal håndteres globalt, og efter at man fx refresher siden. Hvis du har nogle forslag til det er jeg meget lydhør

Reply:
Yes lad os bare køre med det hvis det virker, har ikke arbejdet med global state management i react ;-; 
Ville bare lige lægge en kommentar, men er ikke sikker på hvordan det eventuelt ville skulle gøres anderledes. Jeg skriver det lige på til på tirsdag.